### ISSUE TITLE
Project not running when compiled in single JAR

### ISSUE THREAD

**Issue by ErickPol** [bedatadriven/renjin#418]
We are running a project that incorporates Renjin and it works well when compiled and executed inside an IDE (Netbeans and Intellij) - however, when the project is packaged as a single Jar file the error below is thrown when initializing Renjin. The renjin-core 2696 jar is included in our 'fat jar'.
We have tried versions 2696, 2700 and 2706 (RELEASE). Last week we were able to execute the project without problems : could it be a depedency that has been updated?

**Initializer method:**
```
public class InstanciaRenjin {
    private static InstanciaRenjin instanciaRenjin = null;
    // create a Renjin engine:
    ScriptEngine engine;

    public static InstanciaRenjin getInstance() throws Exception {
        if (instanciaRenjin == null) {
            instanciaRenjin = new InstanciaRenjin();
        }
        return instanciaRenjin;
    }
}
```

**Error message:**
```
java.lang.RuntimeException: java.lang.IllegalStateException: Failed to locate R.home:
        at org.renjin.eval.SessionBuilder.build(SessionBuilder.java:191) ~[renjin-core-0.9.2696.jar!/:na]
        at org.renjin.script.RenjinScriptEngineFactory.getScriptEngine(RenjinScriptEngineFactory.java:110) ~[renjin-script-engine-0.9.2696.jar!/:na]
        at com.opsfactor.instanciaR.InstanciaRenjin.<init>(InstanciaRenjin.java:41) ~[r-instance-0.1.0.jar!/:na]
        at com.opsfactor.instanciaR.InstanciaRenjin.getInstance(InstanciaRenjin.java:33) ~[r-instance-0.1.0.jar!/:na]
        at com.opsfactor.planning.routines.planning.demand.DemandPlanning.geraDemandPlanDFUAgregado(DemandPlanning.java:45) ~[routines-0.1.0.jar!/:na]
        at com.opsfactor.planning.services.planning.DemandPlanningService.geraDemandPlanDFUAgregadoComForecast(DemandPlanningService.java:277) ~[services-0.1.0.jar!/:0.1.0]
        at com.opsfactor.planning.services.planning.front.DemandAnalysisFrontService.getDemandModelSetupDTOFromModelAndProduto(DemandAnalysisFrontService.java:182) ~[services-0.1.0.jar!/:0.1.0]
        at com.opsfactor.planning.web.restcontroller.planningbook.DemandAnalysisRestController.demandAnalysisDTOResponseEntity(DemandAnalysisRestController.java:35) ~[classes!/:0.1.0]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_151]
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[na:1.8.0_151]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:1.8.0_151]
        at java.lang.reflect.Method.invoke(Unknown Source) ~[na:1.8.0_151]
        at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:209) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:136) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:102) [spring-webmvc-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:871) [spring-webmvc-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:777) [spring-webmvc-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) [spring-webmvc-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:991) [spring-webmvc-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:925) [spring-webmvc-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:978) [spring-webmvc-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:870) [spring-webmvc-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:635) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:855) [spring-webmvc-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:742) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) [tomcat-embed-websocket-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177) [spring-security-web-5.0.0.RC1.jar!/:5.0.0.RC1]
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:357) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:270) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:108) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:81) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-5.0.1.RELEASE.jar!/:5.0.1.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:199) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:478) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:803) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1459) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [na:1.8.0_151]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [na:1.8.0_151]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-8.5.23.jar!/:8.5.23]
        at java.lang.Thread.run(Unknown Source) [na:1.8.0_151]
Caused by: java.lang.IllegalStateException: Failed to locate R.home:
        at org.renjin.util.FileSystemUtils.homeDirectoryInCoreJar(FileSystemUtils.java:59) ~[renjin-core-0.9.2696.jar!/:na]
        at org.renjin.eval.Session.<init>(Session.java:143) ~[renjin-core-0.9.2696.jar!/:na]
        at org.renjin.eval.SessionBuilder.build(SessionBuilder.java:179) ~[renjin-core-0.9.2696.jar!/:na]
        ... 90 common frames omitted
Caused by: org.apache.commons.vfs2.FileSystemException: Could not replicate "file:///C:/Users/Erick/Documents/NetBeansProjects/planningSpring/planning-web/target/web-0.1.0.jar!/BOOT-INF/lib/renjin-core-0.9.2696.jar" as it does not exist.
        at org.apache.commons.vfs2.provider.AbstractFileSystem.replicateFile(AbstractFileSystem.java:418) ~[commons-vfs2-2.0.jar!/:2.0]
        at org.renjin.eval.vfs.FastJarFileSystem.<init>(FastJarFileSystem.java:55) ~[renjin-core-0.9.2696.jar!/:na]
        at org.renjin.eval.vfs.FastJarFileProvider.doCreateFileSystem(FastJarFileProvider.java:68) ~[renjin-core-0.9.2696.jar!/:na]
        at org.apache.commons.vfs2.provider.AbstractLayeredFileProvider.createFileSystem(AbstractLayeredFileProvider.java:89) ~[commons-vfs2-2.0.jar!/:2.0]
        at org.apache.commons.vfs2.provider.AbstractLayeredFileProvider.findFile(AbstractLayeredFileProvider.java:63) ~[commons-vfs2-2.0.jar!/:2.0]
        at org.apache.commons.vfs2.impl.DefaultFileSystemManager.resolveFile(DefaultFileSystemManager.java:693) ~[commons-vfs2-2.0.jar!/:2.0]
        at org.apache.commons.vfs2.impl.DefaultFileSystemManager.resolveFile(DefaultFileSystemManager.java:649) ~[commons-vfs2-2.0.jar!/:2.0]
        at org.apache.commons.vfs2.impl.DefaultFileSystemManager.resolveFile(DefaultFileSystemManager.java:605) ~[commons-vfs2-2.0.jar!/:2.0]
        at org.apache.commons.vfs2.provider.res.ResourceFileProvider.findFile(ResourceFileProvider.java:81) ~[commons-vfs2-2.0.jar!/:2.0]
        at org.apache.commons.vfs2.impl.DefaultFileSystemManager.resolveFile(DefaultFileSystemManager.java:693) ~[commons-vfs2-2.0.jar!/:2.0]
        at org.apache.commons.vfs2.impl.DefaultFileSystemManager.resolveFile(DefaultFileSystemManager.java:649) ~[commons-vfs2-2.0.jar!/:2.0]
        at org.apache.commons.vfs2.impl.DefaultFileSystemManager.resolveFile(DefaultFileSystemManager.java:605) ~[commons-vfs2-2.0.jar!/:2.0]
        at org.renjin.util.FileSystemUtils.homeDirectoryInCoreJar(FileSystemUtils.java:50) ~[renjin-core-0.9.2696.jar!/:na]
        ... 92 common frames omitted
```

**Comment by ErickPol** [comment]
Additional insight:
Previous compiled Jars that used to work suddenly started presenting this error.

Here follows 2 logs : the first one generated when the application is run on Netbeans (successfully) and the other from the jar application that does not work anymore:
**Netbeans run:**
```
2018-11-06 11:24:17.451  INFO 20252 --- [nio-8020-exec-3] o.a.c.v.impl.DefaultFileSystemManager    : Using "C:\Users\Erick\AppData\Local\Temp\vfs_cache" as temporary files store.
2018-11-06 11:24:25.320  WARN 20252 --- [nio-8020-exec-3] com.github.fommil.netlib.LAPACK          : Failed to load implementation from: com.github.fommil.netlib.NativeSystemLAPACK
2018-11-06 11:24:44.828  INFO 20252 --- [nio-8020-exec-3] com.github.fommil.jni.JniLoader          : successfully loaded C:\Users\Erick\AppData\Local\Temp\jniloader8031403965355917644netlib-native_ref-win-x86_64.dll
2018-11-06 11:24:47.000  WARN 20252 --- [nio-8020-exec-3] com.github.fommil.netlib.BLAS            : Failed to load implementation from: com.github.fommil.netlib.NativeSystemBLAS
2018-11-06 11:24:47.015  INFO 20252 --- [nio-8020-exec-3] com.github.fommil.jni.JniLoader          : already loaded netlib-native_ref-win-x86_64.dll
```

**Jar run:**
```
2018-11-06 13:42:38.999  INFO 10627 --- [nio-8100-exec-8] o.a.c.v.impl.DefaultFileSystemManager    : Using "/tmp/vfs_cache" as temporary files store.
2018-11-06 13:42:39.002 ERROR 10627 --- [nio-8100-exec-8] c.o.p.w.r.p.DemandAnalysisRestController : null
java.lang.RuntimeException: java.lang.IllegalStateException: Failed to locate R.home:
        at org.renjin.eval.SessionBuilder.build(SessionBuilder.java:191) ~[renjin-core-0.9.2706.jar!/:na]
        at org.renjin.script.RenjinScriptEngineFactory.getScriptEngine(RenjinScriptEngineFactory.java:110) ~[renjin-script-engine-0.9.2706.jar!/:na]
        at com.opsfactor.instanciaR.InstanciaRenjin.<init>(InstanciaRenjin.java:41) ~[r-instance-0.1.0.jar!/:na]
        at com.opsfactor.instanciaR.InstanciaRenjin.getInstance(InstanciaRenjin.java:33) ~[r-instance-0.1.0.jar!/:na]
```

**Comment by akbertram** [comment]
Can you include the code you are using to create your Renjin Session? Are you providing a custom FileSystem or using the default?

**Comment by ErickPol** [comment]
Yes, sure - we are using the default filesystem. I have used the getFileSystemInfo and homeDirectoryInCoreJar methods to gather some log information.
This is running inside a Spring application but this class is an independent singleton so there should be no interference (the same application used to work in the past so I don't believe this is the issue).

**Code:**
```
package com.opsfactor.instanciaR;

import com.opsfactor.planning.support.calendar.Calendario;
import com.opsfactor.planning.support.utility.MetodosUtilidade;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.script.*;
import javax.swing.filechooser.FileSystemView;
import org.apache.commons.vfs2.FileObject;
import org.apache.commons.vfs2.FileSystemException;
import org.apache.commons.vfs2.FileSystemManager;
import org.apache.commons.vfs2.VFS;
import org.renjin.script.*;
import org.renjin.sexp.DoubleVector;
import org.renjin.sexp.StringVector;

/**
 *
 * @author Erick Poletto
 */
public class InstanciaRenjin {
    private static InstanciaRenjin instanciaRenjin = null;
    // create a Renjin engine:
    ScriptEngine engine;

    public static InstanciaRenjin getInstance() throws Exception {
        if (instanciaRenjin == null) {
            instanciaRenjin = new InstanciaRenjin();
        }
        return instanciaRenjin;
    }
    
    private InstanciaRenjin() throws Exception {
        // general file system info
        getFileSystemInfo();
        // debugging vfs call in renjin
        homeDirectoryInCoreJar(VFS.getManager());
        
        RenjinScriptEngineFactory factory = new RenjinScriptEngineFactory();
        // create a Renjin engine:
        engine = factory.getScriptEngine();
        //engine = new Rengine(new String[]{"--no-save"}, false, null);

        // limpa workspace
        engine.eval("rm(list = ls())");
        // importa pacotes
        engine.eval("library(forecast)");
        engine.eval("library(hts)");
        engine.eval("library(rpart)");

        Logger.getLogger(InstanciaRenjin.class.getName()).log(Level.INFO, "R inicializado");
    }
    
    public void getFileSystemInfo() {
        System.out.println("File system roots returned by FileSystemView.getFileSystemView():");
        FileSystemView fsv = FileSystemView.getFileSystemView();
        File[] roots = fsv.getRoots();
        for (int i = 0; i < roots.length; i++)
        {
            System.out.println("Root: " + roots[i]);
        }

        System.out.println("Home directory: " + fsv.getHomeDirectory());

        System.out.println("File system roots returned by File.listRoots():");
        File[] f = File.listRoots();
        for (int i = 0; i < f.length; i++)
        {
            System.out.println("Drive: " + f[i]);
            System.out.println("Display name: " + fsv.getSystemDisplayName(f[i]));
            System.out.println("Is drive: " + fsv.isDrive(f[i]));
            System.out.println("Is floppy: " + fsv.isFloppyDrive(f[i]));
            System.out.println("Readable: " + f[i].canRead());
            System.out.println("Writable: " + f[i].canWrite());
            System.out.println("Total space: " + f[i].getTotalSpace());
            System.out.println("Usable space: " + f[i].getUsableSpace());
        }
    }

  public static String homeDirectoryInCoreJar(FileSystemManager fileSystemManager) {
    // hardcode to the R home location to the classpath location
    // where this class is found.
    System.out.println("getting home directory with VFS FileSystemManager");
    try {
      System.out.println("File System Manager:");
      System.out.println(fileSystemManager.toString());
      String resourceName = "res:org/renjin/sexp/SEXP.class";
      FileObject fileObject = fileSystemManager.resolveFile(resourceName); // <---- where error is thrown
      System.out.println("FileSystem :");
      System.out.println(fileObject.getFileSystem());
      System.out.println("FileObject :");
      System.out.println(fileObject.toString());
      if(!fileObject.exists()) {
        throw new IllegalStateException("Cannot locate resource '" + resourceName + "' in provided virtual file system," +
            " make sure that you are including a ResourceFileProvider with scheme res:");
      }
      System.out.println("FileObject Parent 1:");
      System.out.println(fileObject.getParent().toString());
      System.out.println("FileObject Parent 2:");
      System.out.println(fileObject.getParent().getParent().toString());
      System.out.println("FileObject Parent 2 URL:");
      System.out.println(fileObject.getParent().getParent().getURL().toString());
      
      return fileObject.getParent().getParent().getURL().toString();

    } catch (FileSystemException e) {
      throw new IllegalStateException("Failed to locate R.home: ", e);
    }
  }
}

```
**Output - netbeans execution (runs successfully):**
```
File system roots returned by FileSystemView.getFileSystemView():
Root: C:\Users\Erick\Desktop
Home directory: C:\Users\Erick\Desktop
File system roots returned by File.listRoots():
Drive: C:\
Display name: Acer (C:)
Is drive: true
Is floppy: false
Readable: true
Writable: true
Total space: 999556116480
Usable space: 723242905600
Drive: D:\
Display name: 
Is drive: true
Is floppy: false
Readable: false
Writable: false
Total space: 0
Usable space: 0
getting home directory with VFS FileSystemManager
File System Manager:
org.apache.commons.vfs2.impl.StandardFileSystemManager@1db703de
FileSystem :
org.apache.commons.vfs2.provider.jar.JarFileSystem@418e1d04
FileObject :
jar:file:///C:/Users/Erick/.m2/repository/org/renjin/renjin-core/0.9.2700/renjin-core-0.9.2700.jar!/org/renjin/sexp/SEXP.class
FileObject Parent 1:
jar:file:///C:/Users/Erick/.m2/repository/org/renjin/renjin-core/0.9.2700/renjin-core-0.9.2700.jar!/org/renjin/sexp
FileObject Parent 2:
jar:file:///C:/Users/Erick/.m2/repository/org/renjin/renjin-core/0.9.2700/renjin-core-0.9.2700.jar!/org/renjin
FileObject Parent 2 URL:
jar:file:///C:/Users/Erick/.m2/repository/org/renjin/renjin-core/0.9.2700/renjin-core-0.9.2700.jar!/org/renjin
2018-11-06 18:18:20.597  INFO 10432 --- [nio-8020-exec-1] o.a.c.v.impl.DefaultFileSystemManager    : Using "C:\Users\Erick\AppData\Local\Temp\vfs_cache" as temporary files store.
```

**Output - Jar execution** (error at `FileObject fileObject = fileSystemManager.resolveFile(resourceName);`) . The Renjin jars are inside the fat jar : web-0.1.0.jar\BOOT-INF\lib\renjin-core-0.9.2700.jar.

```
File system roots returned by FileSystemView.getFileSystemView():
Root: C:\Users\Erick\Desktop
Home directory: C:\Users\Erick\Desktop
File system roots returned by File.listRoots():
Drive: C:\
Display name: Acer (C:)
Is drive: true
Is floppy: false
Readable: true
Writable: true
Total space: 999556116480
Usable space: 723245441024
Drive: D:\
Display name:
Is drive: true
Is floppy: false
Readable: false
Writable: false
Total space: 0
Usable space: 0
2018-11-06 18:23:30.781  INFO 25428 --- [nio-8020-exec-4] o.a.c.v.impl.StandardFileSystemManager   : Using "C:\Users\Erick\AppData\Local\Temp\vfs_cache" as temporary files store.
getting home directory with VFS FileSystemManager
File System Manager:
org.apache.commons.vfs2.impl.StandardFileSystemManager@56b2b9b9
2018-11-06 18:23:30.958 ERROR 25428 --- [nio-8020-exec-4] c.o.p.w.r.p.DemandAnalysisRestController : null

java.lang.IllegalStateException: Failed to locate R.home:
        at com.opsfactor.instanciaR.InstanciaRenjin.homeDirectoryInCoreJar(InstanciaRenjin.java:117) ~[r-instance-0.1.0.jar!/:na]
        at com.opsfactor.instanciaR.InstanciaRenjin.<init>(InstanciaRenjin.java:45) ~[r-instance-0.1.0.jar!/:na]
        at com.opsfactor.instanciaR.InstanciaRenjin.getInstance(InstanciaRenjin.java:36) ~[r-instance-0.1.0.jar!/:na]
        at com.opsfactor.planning.routines.planning.demand.DemandPlanning.geraDemandPlanDFUAgregado(DemandPlanning.java:45) ~[routines-0.1.0.jar!/:na]
        at com.opsfactor.planning.services.planning.DemandPlanningService.geraDemandPlanDFUAgregadoComForecast(DemandPlanningService.java:277) ~[services-0.1.0.jar!/:0.1.0]

```

**Comment by akbertram** [comment]
Ok that seems to be in. The Spring Boot ClassLoader returns a URL for SEXP.class that looks like this:

```
jar:file:/home/alex/dev/renjin/core/target/test-classes/org/renjin/eval/vfs/web.jar!/BOOT-INF/lib/renjin.jar!/org/renjin/sexp/SEXP.class
```

ApacheVFS only seems to assume there is only one layer here, and so tries to find a file at the absolute path `/home/alex/dev/renjin/core/target/test-classes/org/renjin/eval/vfs/web.jar!/BOOT-INF/lib/renjin.jar`, which naturally doesn't exist.

If I'm not mistaken, you're getting this error with newer versions of Renjin because we only started checking for a valid R home on initializating with a recent fix (964e65e6598b72aa567c258dcf4d9021625040de).

I will patch up VFS to support this kind of multi-layer URL.

**Comment by akbertram** [comment]
This fix is in release 0.9.2710

**Comment by justinphan3110** [comment]
I still have this problem, my jar stop working when i reach the part using Renjin
